lnrpc/invoicesrpc/addinvoice.go

```diff
@@ -678,169 +627,213 @@ type SelectHopHintsCfg struct {
+// based on the remote balance.
+func getPotentialHints(cfg *SelectHopHintsCfg) ([]*channeldb.OpenChannel,
+	error) {
 
-			return hopHints
-		}
+	// TODO(positiveblue): get the channels slice already filtered by
+	// private == true and sorted by RemoteBalance?
+	openChannels, err := cfg.FetchAllChannels()
+	if err != nil {
+		return nil, err
+	}
 
-		// If this channel can't be a hop hint, then skip it.
-		edgePolicy, canBeHopHint := chanCanBeHopHint(channel, cfg)
-		if edgePolicy == nil || !canBeHopHint {
-			continue
+	privateChannels := make([]*channeldb.OpenChannel, 0, len(openChannels))
+	for _, oc := range openChannels {
+		isPublic := oc.ChannelFlags&lnwire.FFAnnounceChannel != 0
+		if !isPublic {
+			privateChannels = append(privateChannels, oc)
 		}
+	}
 
-		// Similarly, in this first pass, we'll ignore all channels in
-		// isolation can't satisfy this payment.
-		if channel.RemoteBalance < amtMSat {
-			continue
-		}
+	// Sort the channels in descending remote balance.
+	compareRemoteBalance := func(i, j int) bool {
+		iBalance := privateChannels[i].LocalCommitment.RemoteBalance
+		jBalance := privateChannels[j].LocalCommitment.RemoteBalance
+		return iBalance > jBalance
+	}
+	sort.Slice(privateChannels, compareRemoteBalance)
 
-		// Lookup and see if there is an alias SCID that exists.
-		chanID := lnwire.NewChanIDFromOutPoint(
-			&channel.FundingOutpoint,
-		)
-		alias, _ := cfg.GetAlias(chanID)
-
-		// If this is a channel where the option-scid-alias feature bit
-		// was negotiated and the alias is not yet assigned, we cannot
-		// issue an invoice. Doing so might expose the confirmed SCID
-		// of a private channel.
-		if channel.ScidAliasFeature {
-			var defaultScid lnwire.ShortChannelID
-			if alias == defaultScid {
-				continue
-			}
+	return privateChannels, nil
+}
+
+// shouldIncludeChannel returns true if the channel passes all the checks to
+// be a hopHint in a given invoice.
+func shouldIncludeChannel(channel *channeldb.OpenChannel,
+	alreadyIncluded map[uint64]struct{},
+	cfg *SelectHopHintsCfg) (zpay32.HopHint, lnwire.MilliSatoshi, bool) {
+
+	if _, ok := alreadyIncluded[channel.ShortChannelID.ToUint64()]; ok {
+		return zpay32.HopHint{}, 0, false
+	}
+
+	chanID := lnwire.NewChanIDFromOutPoint(
+		&channel.FundingOutpoint,
+	)
+
+	hopHintInfo := newHopHintInfo(channel, cfg.IsChannelActive(chanID))
+
+	// If this channel can't be a hop hint, then skip it.
+	edgePolicy, canBeHopHint := chanCanBeHopHint(hopHintInfo, cfg)
+	if edgePolicy == nil || !canBeHopHint {
+		return zpay32.HopHint{}, 0, false
+	}
+
+	if hopHintInfo.ScidAliasFeature {
+		alias, err := cfg.GetAlias(chanID)
+		if err != nil {
+			return zpay32.HopHint{}, 0, false
 		}
 
-		// Now that we now this channel use usable, add it as a hop
-		// hint and the indexes we'll use later.
-		addHopHint(&hopHints, channel, edgePolicy, alias)
+		_, included := alreadyIncluded[alias.ToUint64()]
+		switch {
+		case alias.IsDefault():
```

Comment:
I think we can remove the switch statement entirely. The first case is not at all dependent on if the alias is already included or not. 
So I would suggest:
```
if alias.IsDefault() || alreadyIncluded[alias.ToUint64()] {
    return zpay32.HopHint{}, 0, false
}

hopHintInfo.ShortChannelID = alias.ToUint64()
```